In Sys.getcwd, don't call getwd if HAS_GETWD is not set (follow-up to #1086)

xavierleroy opened the pull request:
This is a proper fix for the bug reported in #1086 but incorrectly fixed there.

xavierleroy committed:
GPR#1086: in Sys.getcwd, don't call getwd if HAS_GETWD is not set

yallop commented on byterun/sys.c:
```diff
@@ -336,11 +336,13 @@ CAMLprim value caml_sys_chdir(value dirname)
 CAMLprim value caml_sys_getcwd(value unit)
 {
   char buff[4096];
```
This is perhaps slightly off-topic, but shouldn't the length of `buff` use `PATH_MAX` rather than `4096`?

xavierleroy replied:
Well, for modern OSes and file systems, there is no fixed limit on the length of an absolute file path, so 4096 is not much more wrong than `PATH_MAX`...

We probably could do a better job by dynamically resizing the buffer if `getcwd` errors with `ERANGE`.  But that's a topic for another pull request.

alainfrisch commented:
Copying a comment from @gasche on the other PR:

> there is another occurrence of getwd in asmrun/spacetime.c that you may want to change as well

shindere commented:
Xavier Leroy (2017/09/14 18:00 +0000):
> xavierleroy commented on this pull request.
> 
> 
> 
> > @@ -336,11 +336,13 @@ CAMLprim value caml_sys_chdir(value dirname)
>  CAMLprim value caml_sys_getcwd(value unit)
>  {
>    char buff[4096];
> 
> Well, for modern OSes and file systems, there is no fixed limit on the
> length of an absolute file path, so 4096 is not much more wrong than
> `PATH_MAX`...

Perhaps not much more wrong, but it carries less semeantics.

Reminds me the quote saying that programs are written mainly to be read,
not to be compiled. :)

damiendoligez commented:
cc @mshinwell on Gabriel's note about `spacetime.c`.

mshinwell committed:
Fix Spacetime for HAS_GETCWD / HAS_GETWD

mshinwell commented:
I believe I have fixed `spacetime.c` directly on this branch.
I also merged with current trunk.

The `PATH_MAX` issue is a potential security concern I think (depending on the no doubt many and various implementations of `getwd`).

xavierleroy committed:
GPR#1086 continued: do not use getwd()

The C library function getwd() is risky to use because of the implicit length of the buffer.  Moreover it was marked as deprecated in POSIX 2001 and is gone in POSIX 2008.  Everyone should have getcwd() nowadays.  The Microsoft C runtime has it.  Let's use it.

xavierleroy commented:
@mshinwell is correct that `getwd` is a security risk.  So, let's stop using it entirely  (commit [01c90f5]).  Quoting the man page:
```
       getwd() is present in POSIX.1-2001, but  marked  LEGACY.   POSIX.1-2008
       removes   the   specification   of   getwd().   Use  getcwd()  instead.
       POSIX.1-2001 does not define any errors for getwd().
```
Even the Microsoft CRT provides `getcwd`, so that should be good enough for everyone.

I stand by my decision to not use `PATH_MAX`.  Quoting POSIX.1-2008:
> A definition of one of the symbolic constants in the following list [including `PATH_MAX`] shall be omitted from the <limits.h> header on specific implementations where the corresponding value is equal to or greater than the stated minimum, but where the value can vary depending on the file to which it is applied. The actual value supported for a specific pathname shall be provided by the pathconf() function.

So, not only `PATH_MAX` may not be a constant, but it may not be defined at all.  Enough.

damiendoligez commented:
@mshinwell you should also avoid `PATH_MAX` in the spacetime code.

xavierleroy committed:
GPR#1086 continued: do not use PATH_MAX to size the buffer for getcwd()

According to POSIX 2008.1, PATH_MAX may not be a constant, and even worse, it may not be defined at all.  Better go back to a fixed-size buffer for now.  Later we could wrap getcwd() in a function that resizes the buffer until it's big enough.

xavierleroy commented:
I took the liberty of reverting the Spacetime code to use a fixed-size buffer and forget about PATH_MAX.  I also rebased on current trunk.  Requesting final review and permission to merge.

mshinwell commented on Changes:
```diff
@@ -609,6 +609,9 @@ Release branch for 4.06:
   (Stephen Dolan)
 
+- GPR#1086: in Sys.getcwd, don't call getwd() if HAS_GETWD is not set
+  (report and first fix by Sebastian Markbåge, final fix by Xavier Leroy)
```
This needs updating to reflect the fact that we no longer use `getwd` at all.

xavierleroy replied:
Done.

mshinwell commented on asmrun/spacetime.c:
```diff
@@ -225,17 +225,15 @@ void caml_spacetime_initialize(void)
 #endif
         if (dir_ok) {
-          automatic_snapshot_dir = strdup(cwd);
+          automatic_snapshot_dir = cwd;
```
This is wrong, we still need `strdup`, since `cwd` is on the stack.

xavierleroy replied:
Well spotted, sorry about that.  Fixed.

mshinwell commented:
OK to merge after the two changes I just pointed out.

xavierleroy committed:
GPR#1086 continued: finish reverting to using a fixed-size getcwd buffer

The buffer, being allocated on the stack (again), needs copying (again).

xavierleroy committed:
GPR#1086 continued: do not use getwd()

mshinwell commented:
Merged to trunk and cherry-picked to 4.06 (`538e110ad`).
